File: problems/src/clojure/src/ch1_arrays_and_strings/ch1_2.clj
(ns ch1-arrays-and-strings.ch1-2)

; Reverse a string
; Just apply the str function to each of the seqs of chars resulting from reverse
(apply str (reverse "abc"))
(apply str (reverse nil))
(apply str (reverse "a"))

; What it boils down to:
(apply str (reduce conj () "abc"))
; Because seqs add to the front, iterating over the word and putting each element as
; the head of the return seq will reverse the word.
; (apply str (reduce conj [] "abc")) => "abc"
; Vectors add at the tail, so it just copies exactly

; Reduce is interesting here. It's

File: problems/src/java/src/ch1_arrays_and_strings/Ch1_2.java
package ch1_arrays_and_strings;

public class Ch1_2 {

    public static void main(String[] args) {
        System.out.println(reverse(""));
        System.out.println(reverse(null));
        System.out.println(reverse("a"));
        System.out.println(reverse("as"));
        System.out.println(reverse("asdf"));
        System.out.println(reverse("asdfe"));
    }

    public static String reverse(String str) {
        if (str == null) return null;
        char[] chars = str.toCharArray();

        int head = 0;
        int tail = chars.length - 1;
        while (head < tail) {
            char temp = chars[head];
            chars[head] = chars[tail];
            chars[tail] = temp;
            head++;
            tail--;
        }

        return new String(chars);
    }
}
